Eslint errors when using eslint-loader with version beta.7

It appears version 0.4.0-beta.7 is giving me a slew of eslint errors through eslint-loader. I have `enforce: 'pre'` set for eslint-loader to ensure it is executed on my untouched source files prior to babel-loader, but it seems like this plugin does not respect that.

I did not have this problem with version 0.4.0-beta.5 and everything was working great. Is there a new configuration that I am missing?

Let me know if you need any additional info.

Thanks for all your hard work!

**Example output:**
```
 error  in ./src/app/views/public/views/sponsor/index.js

Module Error (from ./node_modules/eslint-loader/dist/cjs.js):

/home/corey/workspace/seeitmove/src/app/views/public/views/sponsor/index.js
   1:1   error  '$RefreshRuntime$' is not defined                                                                                                                                                               no-undef
   2:1   error  '$RefreshSetup$' is not defined                                                                                                                                                                 no-undef
   4:1   error  Import in body of module; reorder to top                                                                                                                                                        import/first
   5:1   error  Import in body of module; reorder to top                                                                                                                                                        import/first
  17:1   error  Delete `⏎`                                                                                                                                                                                      prettier/prettier
  18:24  error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  19:1   error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  19:56  error  Replace `currentExports,·module.id` with `⏎··currentExports,⏎··module.id,⏎`                                                                                                                     prettier/prettier
  25:7   error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  36:8   error  Insert `,`                                                                                                                                                                                      prettier/prettier
  46:18  error  Identifier '__react_refresh_error_overlay__' is not in camel case                                                                                                                               camelcase
  47:11  error  Identifier '__react_refresh_error_overlay__' is not in camel case                                                                                                                               camelcase
  47:11  error  '__react_refresh_error_overlay__' is not defined                                                                                                                                                no-undef
  49:11  error  '__react_refresh_error_overlay__' is not defined                                                                                                                                                no-undef
  52:13  error  Replace `typeof·__react_refresh_test__·!==·'undefined'·&&·__react_refresh_test__` with `⏎··········typeof·__react_refresh_test__·!==·'undefined'·&&⏎··········__react_refresh_test__⏎········`  prettier/prettier
  52:20  error  Identifier '__react_refresh_test__' is not in camel case                                                                                                                                        camelcase
  52:62  error  Identifier '__react_refresh_test__' is not in camel case                                                                                                                                        camelcase
  52:62  error  '__react_refresh_test__' is not defined                                                                                                                                                         no-undef
  58:9   error  '__webpack_require__' is not defined                                                                                                                                                            no-undef
  59:8   error  Insert `,`                                                                                                                                                                                      prettier/prettier
  64:9   error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  65:9   error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  65:70  error  Replace `prevExports,·currentExports` with `⏎··········prevExports,⏎··········currentExports,⏎········`                                                                                         prettier/prettier
  69:9   error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  76:22  error  Identifier '__react_refresh_error_overlay__' is not in camel case                                                                                                                               camelcase
  77:15  error  Identifier '__react_refresh_error_overlay__' is not in camel case                                                                                                                               camelcase
  77:15  error  '__react_refresh_error_overlay__' is not defined                                                                                                                                                no-undef
  79:15  error  '__react_refresh_error_overlay__' is not defined                                                                                                                                                no-undef
  81:12  error  Insert `,`                                                                                                                                                                                      prettier/prettier
  86:9   error  Replace `isHotUpdate·&&·__react_refresh_utils__.isReactRefreshBoundary(prevExports)` with `⏎······isHotUpdate·&&⏎······__react_refresh_utils__.isReactRefreshBoundary(prevExports)⏎····`        prettier/prettier
  86:24  error  '__react_refresh_utils__' is not defined                                                                                                                                                        no-undef
  90:2   error  Insert `⏎`                                                                                                                                                                                      prettier/prettier

✖ 32 problems (32 errors, 0 warnings)
  11 errors and 0 warnings potentially fixable with the `--fix` option.


 @ ./src/app/views/public/index.js 12:0-43 26:46-58
 @ ./src/app/index.js
 @ ./internals/entries/client.js
 @ multi ./node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ReactRefreshEntry.js ./node_modules/@pmmmwh/react-refresh-webpack-plugin/client/ErrorOverlayEntry.js?sockPort=3001 ./internals/entries/client.js
```

**Relevant Webpack Config:**
```javascript
import ReactRefreshPlugin from '@pmmmwh/react-refresh-webpack-plugin';
import babelConfig from '../../.babelrc';

export default {
  module: {
    rules: [
      {
        exclude: /node_modules/,
        test: /\.js$/,
        loader: 'babel-loader',
        options: {
          babelrc: false,
          presets: babelConfig.presets,
          plugins: [
            ...babelConfig.plugins,
            ...(target === 'web'
              ? [require.resolve('react-refresh/babel')]
              : []),
          ],
        },
      },
      {
        enforce: 'pre',
        test: /\.js$/,
        loader: 'eslint-loader',
        options: {
          cache: true,
        },
      },
    ],
  },
  plugins: [
    new ReactRefreshPlugin({
      overlay: {
        sockPort: 3001,
      },
    }),
  ],
}
```

--- a/lib/index.js
+++ b/lib/index.js
@@ -185,10 +185,11 @@ class ReactRefreshPlugin {
                 }
 
                 const moduleInterceptor = Template.asString([
+                  `${refreshGlobal}.init();`,
                   'try {',
                   Template.indent(lines[moduleInitializationLineNumber]),
                   '} finally {',
-                  Template.indent(`${refreshGlobal}.cleanup();`),
+                  Template.indent(`${refreshGlobal}.cleanup(moduleId);`),
                   '}',
                 ]);
 

--- a/lib/runtime/RefreshRuntimeModule.js
+++ b/lib/runtime/RefreshRuntimeModule.js
@@ -22,12 +22,13 @@ class ReactRefreshRuntimeModule extends RuntimeModule {
         `options.factory = ${runtimeTemplate.basicFunction(
           'moduleObject, moduleExports, webpackRequire',
           [
+            `${refreshGlobal}.init();`,
             'try {',
             Template.indent(
               'originalFactory.call(this, moduleObject, moduleExports, webpackRequire);'
             ),
             '} finally {',
-            Template.indent(`${refreshGlobal}.cleanup();`),
+            Template.indent(`${refreshGlobal}.cleanup(options.id);`),
             '}',
           ]
         )}`,

--- a/lib/utils/getRefreshGlobal.js
+++ b/lib/utils/getRefreshGlobal.js
@@ -31,30 +31,36 @@ function getRefreshGlobal(runtimeTemplate = FALLBACK_RUNTIME_TEMPLATE) {
   return Template.asString([
     `${refreshGlobal} = {`,
     Template.indent([
-      `cleanup: ${runtimeTemplate.returningFunction('undefined')},`,
-      `register: ${runtimeTemplate.returningFunction('undefined')},`,
-      'runtime: {},',
+      `init: ${runtimeTemplate.basicFunction('', [
+        `this.cleanup = ${runtimeTemplate.returningFunction('undefined')};`,
+        `this.register = ${runtimeTemplate.returningFunction('undefined')};`,
+        'this.runtime = {};',
+        `this.signature = ${runtimeTemplate.returningFunction(
+          runtimeTemplate.returningFunction('type', 'type')
+        )};`,
+      ])}`,
       `setup: ${runtimeTemplate.basicFunction('currentModuleId', [
-        `${declaration} prevCleanup = ${refreshGlobal}.cleanup;`,
-        `${declaration} prevReg = ${refreshGlobal}.register;`,
-        `${declaration} prevSig = ${refreshGlobal}.signature;`,
+        `${declaration} prevCleanup = this.cleanup;`,
+        `${declaration} prevReg = this.register;`,
+        `${declaration} prevSig = this.signature;`,
         '',
-        `${refreshGlobal}.register = ${runtimeTemplate.basicFunction('type, id', [
+        `this.register = ${runtimeTemplate.basicFunction('type, id', [
           `${declaration} typeId = currentModuleId + " " + id;`,
-          `${refreshGlobal}.runtime.register(type, typeId);`,
+          'this.runtime.register(type, typeId);',
         ])}`,
         '',
-        `${refreshGlobal}.signature = ${refreshGlobal}.runtime.createSignatureFunctionForTransform;`,
+        'this.signature = this.runtime.createSignatureFunctionForTransform;',
         '',
-        `${refreshGlobal}.cleanup = ${runtimeTemplate.basicFunction('', [
-          `${refreshGlobal}.register = prevReg;`,
-          `${refreshGlobal}.signature = prevSig;`,
-          `${refreshGlobal}.cleanup = prevCleanup;`,
+        `this.cleanup = ${runtimeTemplate.basicFunction('cleanupModuleId', [
+          'if (currentModuleId === cleanupModuleId) {',
+          Template.indent([
+            'this.register = prevReg;',
+            'this.signature = prevSig;',
+            'this.cleanup = prevCleanup;',
+          ]),
+          '}',
         ])}`,
       ])},`,
-      `signature: ${runtimeTemplate.returningFunction(
-        runtimeTemplate.returningFunction('type', 'type')
-      )}`,
     ]),
     '};',
   ]);

--- a/lib/utils/getRefreshGlobal.js
+++ b/lib/utils/getRefreshGlobal.js
@@ -32,31 +32,31 @@ function getRefreshGlobal(runtimeTemplate = FALLBACK_RUNTIME_TEMPLATE) {
     `${refreshGlobal} = {`,
     Template.indent([
       `init: ${runtimeTemplate.basicFunction('', [
-        `this.cleanup = ${runtimeTemplate.returningFunction('undefined')};`,
-        `this.register = ${runtimeTemplate.returningFunction('undefined')};`,
-        'this.runtime = {};',
-        `this.signature = ${runtimeTemplate.returningFunction(
+        `${refreshGlobal}.cleanup = ${runtimeTemplate.returningFunction('undefined')};`,
+        `${refreshGlobal}.register = ${runtimeTemplate.returningFunction('undefined')};`,
+        `${refreshGlobal}.runtime = {};`,
+        `${refreshGlobal}.signature = ${runtimeTemplate.returningFunction(
           runtimeTemplate.returningFunction('type', 'type')
         )};`,
-      ])}`,
+      ])},`,
       `setup: ${runtimeTemplate.basicFunction('currentModuleId', [
-        `${declaration} prevCleanup = this.cleanup;`,
-        `${declaration} prevReg = this.register;`,
-        `${declaration} prevSig = this.signature;`,
+        `${declaration} prevCleanup = ${refreshGlobal}.cleanup;`,
+        `${declaration} prevReg = ${refreshGlobal}.register;`,
+        `${declaration} prevSig = ${refreshGlobal}.signature;`,
         '',
-        `this.register = ${runtimeTemplate.basicFunction('type, id', [
+        `${refreshGlobal}.register = ${runtimeTemplate.basicFunction('type, id', [
           `${declaration} typeId = currentModuleId + " " + id;`,
-          'this.runtime.register(type, typeId);',
+          `${refreshGlobal}.runtime.register(type, typeId);`,
         ])}`,
         '',
-        'this.signature = this.runtime.createSignatureFunctionForTransform;',
+        `${refreshGlobal}.signature = ${refreshGlobal}.runtime.createSignatureFunctionForTransform;`,
         '',
-        `this.cleanup = ${runtimeTemplate.basicFunction('cleanupModuleId', [
+        `${refreshGlobal}.cleanup = ${runtimeTemplate.basicFunction('cleanupModuleId', [
           'if (currentModuleId === cleanupModuleId) {',
           Template.indent([
-            'this.register = prevReg;',
-            'this.signature = prevSig;',
-            'this.cleanup = prevCleanup;',
+            `${refreshGlobal}.register = prevReg;`,
+            `${refreshGlobal}.signature = prevSig;`,
+            `${refreshGlobal}.cleanup = prevCleanup;`,
           ]),
           '}',
         ])}`,

--- a/lib/utils/injectRefreshLoader.js
+++ b/lib/utils/injectRefreshLoader.js
@@ -24,15 +24,11 @@ function injectRefreshLoader(data, matchObject) {
     // Check to prevent double injection
     !data.loaders.find(({ loader }) => loader === resolvedLoader)
   ) {
-    // It is important to have inject loader before anything,
-    // especially if runtime transforms (e.g. @babel/plugin-transform-runtime) are being used.
-    // Runtime transforms might make use of cache mechanisms (e.g. WeakMaps) to handle imports,
-    // so if we do things out of their loop,
-    // we might risk referencing a different version of the same dependency in the runtime.
-    // This in its worst form will break HMR referential equality checks,
-    // making us bail out on every change.
-    // e.g. React in the file will not be equal to React as consumed by react-refresh.
-    data.loaders.push({
+    // As we inject runtime code for each module,
+    // it is important to run the injected loader after everything.
+    // This way we can ensure that all code-processing have been done,
+    // and we won't risk breaking tools like Flow or ESLint.
+    data.loaders.unshift({
       loader: resolvedLoader,
       options: undefined,
     });

--- a/test/unit/getRefreshGlobal.test.js
+++ b/test/unit/getRefreshGlobal.test.js
@@ -9,13 +9,16 @@ describe('getRefreshGlobal', () => {
     delete global.__webpack_require__;
   });
 
-  it('should return refresh global template without providing runtime template', () => {
-    const refreshGlobal = getRefreshGlobal();
-    expect(refreshGlobal).toMatchInlineSnapshot(`
+  it('should return working refresh global without providing runtime template', () => {
+    const refreshGlobalTemplate = getRefreshGlobal();
+    expect(refreshGlobalTemplate).toMatchInlineSnapshot(`
       "__webpack_require__.$Refresh$ = {
-      	cleanup: function() { return undefined; },
-      	register: function() { return undefined; },
-      	runtime: {},
+      	init: function() {
+      		__webpack_require__.$Refresh$.cleanup = function() { return undefined; };
+      		__webpack_require__.$Refresh$.register = function() { return undefined; };
+      		__webpack_require__.$Refresh$.runtime = {};
+      		__webpack_require__.$Refresh$.signature = function() { return function(type) { return type; }; };
+      	},
       	setup: function(currentModuleId) {
       		var prevCleanup = __webpack_require__.$Refresh$.cleanup;
       		var prevReg = __webpack_require__.$Refresh$.register;
@@ -28,57 +31,80 @@ describe('getRefreshGlobal', () => {
 
       		__webpack_require__.$Refresh$.signature = __webpack_require__.$Refresh$.runtime.createSignatureFunctionForTransform;
 
-      		__webpack_require__.$Refresh$.cleanup = function() {
-      			__webpack_require__.$Refresh$.register = prevReg;
-      			__webpack_require__.$Refresh$.signature = prevSig;
-      			__webpack_require__.$Refresh$.cleanup = prevCleanup;
+      		__webpack_require__.$Refresh$.cleanup = function(cleanupModuleId) {
+      			if (currentModuleId === cleanupModuleId) {
+      				__webpack_require__.$Refresh$.register = prevReg;
+      				__webpack_require__.$Refresh$.signature = prevSig;
+      				__webpack_require__.$Refresh$.cleanup = prevCleanup;
+      			}
       		}
       	},
-      	signature: function() { return function(type) { return type; }; }
       };"
     `);
     expect(() => {
-      eval(refreshGlobal);
+      eval(refreshGlobalTemplate);
     }).not.toThrow();
+
+    const refreshGlobal = global.__webpack_require__.$Refresh$;
+    expect(() => {
+      refreshGlobal.init();
+    }).not.toThrow();
+    expect(typeof refreshGlobal.cleanup).toBe('function');
+    expect(typeof refreshGlobal.register).toBe('function');
+    expect(typeof refreshGlobal.signature).toBe('function');
+    expect(typeof refreshGlobal.runtime).toBe('object');
   });
 
   it.skipIf(
     WEBPACK_VERSION !== 5,
-    'should return refresh global template with provided runtime template',
+    'should return working refresh global with provided runtime template',
     () => {
       const RuntimeTemplate = require('webpack/lib/RuntimeTemplate');
-      const refreshGlobal = getRefreshGlobal(
+      const refreshGlobalTemplate = getRefreshGlobal(
         new RuntimeTemplate({ ecmaVersion: 6 }, { shorten: (item) => item })
       );
-      expect(refreshGlobal).toMatchInlineSnapshot(`
-      "__webpack_require__.$Refresh$ = {
-      	cleanup: () => undefined,
-      	register: () => undefined,
-      	runtime: {},
-      	setup: (currentModuleId) => {
-      		const prevCleanup = __webpack_require__.$Refresh$.cleanup;
-      		const prevReg = __webpack_require__.$Refresh$.register;
-      		const prevSig = __webpack_require__.$Refresh$.signature;
+      expect(refreshGlobalTemplate).toMatchInlineSnapshot(`
+        "__webpack_require__.$Refresh$ = {
+        	init: () => {
+        		__webpack_require__.$Refresh$.cleanup = () => undefined;
+        		__webpack_require__.$Refresh$.register = () => undefined;
+        		__webpack_require__.$Refresh$.runtime = {};
+        		__webpack_require__.$Refresh$.signature = () => (type) => type;
+        	},
+        	setup: (currentModuleId) => {
+        		const prevCleanup = __webpack_require__.$Refresh$.cleanup;
+        		const prevReg = __webpack_require__.$Refresh$.register;
+        		const prevSig = __webpack_require__.$Refresh$.signature;
 
-      		__webpack_require__.$Refresh$.register = (type, id) => {
-      			const typeId = currentModuleId + \\" \\" + id;
-      			__webpack_require__.$Refresh$.runtime.register(type, typeId);
-      		}
+        		__webpack_require__.$Refresh$.register = (type, id) => {
+        			const typeId = currentModuleId + \\" \\" + id;
+        			__webpack_require__.$Refresh$.runtime.register(type, typeId);
+        		}
 
-      		__webpack_require__.$Refresh$.signature = __webpack_require__.$Refresh$.runtime.createSignatureFunctionForTransform;
+        		__webpack_require__.$Refresh$.signature = __webpack_require__.$Refresh$.runtime.createSignatureFunctionForTransform;
 
-      		__webpack_require__.$Refresh$.cleanup = () => {
-      			__webpack_require__.$Refresh$.register = prevReg;
-      			__webpack_require__.$Refresh$.signature = prevSig;
-      			__webpack_require__.$Refresh$.cleanup = prevCleanup;
-      		}
-      	},
-      	signature: () => (type) => type
-      };"
-    `);
+        		__webpack_require__.$Refresh$.cleanup = (cleanupModuleId) => {
+        			if (currentModuleId === cleanupModuleId) {
+        				__webpack_require__.$Refresh$.register = prevReg;
+        				__webpack_require__.$Refresh$.signature = prevSig;
+        				__webpack_require__.$Refresh$.cleanup = prevCleanup;
+        			}
+        		}
+        	},
+        };"
+      `);
       expect(() => {
-        eval(refreshGlobal);
+        eval(refreshGlobalTemplate);
       }).not.toThrow();
+
+      const refreshGlobal = global.__webpack_require__.$Refresh$;
+      expect(() => {
+        refreshGlobal.init();
+      }).not.toThrow();
+      expect(typeof refreshGlobal.cleanup).toBe('function');
+      expect(typeof refreshGlobal.register).toBe('function');
+      expect(typeof refreshGlobal.signature).toBe('function');
+      expect(typeof refreshGlobal.runtime).toBe('object');
     }
   );
 });

